feat: allow for multiple external provider specs

cdoern opened the pull request:
# What does this PR do?

when using the providers.d method of installation users could hand craft their AdapterSpec's to use overlapping code meaning one repo could contain an inline and remote impl. Currently installing a provider via module does not allow for that as each repo is only allowed to have one `get_provider_spec` method with one Spec returned

add an optional way for `get_provider_spec` to return a list of `ProviderSpec` where each can be either an inline or remote impl.

Note: the `adapter_type` in `get_provider_spec` MUST match the `provider_type` in the build/run yaml for this to work.

resolves #3226

## Test Plan

once this merges we need to re-enable the external provider test and account for this functionality. Work needs to be done in the external provider repos to support this functionality.

ashwinb commented on llama_stack/core/distribution.py:
```diff
@@ -264,9 +265,16 @@ def get_external_providers_from_module(
-                # in the case we are building we CANNOT import this module of course because it has not been installed.
-                # return a partially filled out spec that the build script will populate.
-                registry[Api(provider_api)][provider_type] = spec
+                if isinstance(spec, list):
```
where is the corresponding type definition change which changes the type into a `ProviderSpec | list[ProviderSpec]`?

ashwinb commented on llama_stack/core/distribution.py:
```diff
@@ -264,9 +265,16 @@ def get_external_providers_from_module(
+                    # optionally allow people to pass inline and remote provider specs as a returned list.
+                    # with the old method, users could pass in directories of specs using overlapping code
+                    # we want to ensure we preserve that flexibility in this method.
+                    for adapter in spec:
```
can you have three adapters? what if there is _no_ adapter? adapter implies remote right?

cdoern replied:
but I guess since they both inherit from `ProviderSpec`, I can depend just on `provider_type` .

leseb commented on llama_stack/core/distribution.py:
```diff
@@ -264,9 +265,16 @@ def get_external_providers_from_module(
+                    # optionally allow people to pass inline and remote provider specs as a returned list.
+                    # with the old method, users could pass in directories of specs using overlapping code
+                    # we want to ensure we preserve that flexibility in this method.
+                    for provider_spec in spec:
+                        if provider_spec.provider_type != provider.provider_type:
+                            continue
+                        registry[Api(provider_api)][provider.provider_type] = provider_spec
```
I'd like to add more logging in here, otherwise, I'm fine with the approach.

cdoern commented:
changes:

I added `TestGetExternalProvidersFromModule` to `test_distribution.py` and put basically every testI could think of in there for expected outputs, edge cases, etc.

I added two log statements to `get_external_providers_from_module` indicating when multiple specs are found from an external provider's `get_provider_spec`
